### PR TITLE
API-5838 - LOAST accepts bearer token security

### PR DESCRIPTION
**Pull request by alexandervalencia**
This PR supports [API-5838](https://vajira.max.gov/browse/API-5838)

LOAST should now prompt the user for their bearer token if the OAS uses an `http` type security with the scheme of `bearer`.

### CODE REVIEW

File: src/commands/positive.ts [hunk 1]
```diff
@@ -192,18 +198,37 @@ export default class Positive extends Command {
...
       }
     }
 
     if (securityTypes[OASSecurityType.APIKEY]) {
-      const apiSecurityName = securityTypes[OASSecurityType.APIKEY];
+      const apiSecurityName = securityTypes[OASSecurityType.APIKEY].key;
       const value =
         flags.apiKey ??
-        (await cli.prompt('What is your API Key?', { type: 'mask' }));
+        (await cli.prompt('Please provide your API Key?', { type: 'mask' }));
```
Comment (michaelkimball) [inline src/commands/positive.ts:213]: Need to remove the `?` if you change it to a statement.

File: test/commands/positive.test.ts [hunk 2]
```diff
@@ -107,7 +107,7 @@ describe('Positive', () => {
             },
           ],
         },
-        [{ 'boromir-security': [] }],
+        [{ 'faramir-security': [] }],
```
Comment (michaelkimball) [inline test/commands/positive.test.ts:110]: More Faramir appreciation 🎉

File: test/fixtures/securities/bearer_token_oas.json [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+      "haladir-security": {
+        "type": "http",
+        "scheme": "bearer",
+        "bearerFormat": "JWT"
+      }
+    }
+  },
+  "security": [
+    {
+      "haladir-security": []
```
Comment (michaelkimball) [inline test/fixtures/securities/bearer_token_oas.json:226]: haldir

File: test/fixtures/securities/bearer_token_oas.json [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+          "401": {
+            "description": "Unauthorized Request"
+          },
+          "403": {
+            "description": "Bad API Token"
+          }
+        },
+        "security": [
+          {
+            "faramir_api_key": []
```
Comment (michaelkimball) [inline test/fixtures/securities/bearer_token_oas.json:73]: Does this complain? I feel like this should complain

File: test/utilities/oas-operation/oas-operation.test.ts [hunk 5]
```diff
@@ -164,4 +167,20 @@ describe('OASOperation', () => {
...
+      ]);
+
+      expect(secureOperation.security).toEqual([
+        { key: 'boromir-security', scopes: undefined },
+      ]);
+    });
+
+    it('overrides spec level security with operation level security if it exists', () => {
+      expect(operation.security).toEqual([]);
+    });
```
Comment (christineyost) [inline test/utilities/oas-operation/oas-operation.test.ts:184]: Not sure I understand this test

File: test/utilities/oas-operation/oas-operation.test.ts [hunk 6]
```diff
@@ -164,4 +167,20 @@ describe('OASOperation', () => {
       expect(operation.getResponseSchema(683)).toBeNull();
     });
   });
+
+  describe('getSecurity', () => {
+    it('uses operation level security if it exists', () => {
```
Comment (christineyost) [inline test/utilities/oas-operation/oas-operation.test.ts:172]: Not sure if this suggestion is accurate

```suggestion
    it('returns operation level security if it exists', () => {
```

File: src/commands/positive.ts [hunk 7]
```diff
@@ -192,18 +198,37 @@ export default class Positive extends Command {
...
 
       this.securityValues[apiSecurityName] = { value };
     }
+
+    if (securityTypes[OASSecurityType.HTTP]) {
+      const bearerSecurityName = securityTypes[OASSecurityType.HTTP].key;
+      let value;
+
+      if (securityTypes[OASSecurityType.HTTP].scheme === 'bearer') {
```
Comment (christineyost) [inline src/commands/positive.ts:222]: Might be nice to make 'bearer' a constant. We have one other defined in the `src/utilities/constants.ts` file

**Review by christineyost — COMMENTED** [review]
Looks good to me! Added a few small thoughts.

I was also just thinking that if we change the way we create security schemes in Docserver we'll need to do some work to update LOAST as well. I'm not sure if we're close to a decision on that though. Might be best to get this in and re-address if we end up updating Docserver?

File: src/commands/positive.ts (hunk 7 above)
Comment (alexandervalencia) [reply src/commands/positive.ts:222]: Great suggestion!

File: test/utilities/oas-operation/oas-operation.test.ts (hunk 5 above)
Comment (alexandervalencia) [reply test/utilities/oas-operation/oas-operation.test.ts:184]: I think I changed the test and forgot to update the description, I'll check it out

File: test/utilities/oas-operation/oas-operation.test.ts (hunk 6 above)
Comment (alexandervalencia) [reply test/utilities/oas-operation/oas-operation.test.ts:172]: fixed this whole section. thanks!

**Comment by alexandervalencia** [comment]
I was able to successfully verify that `bearer` security is working as expected:
```
./bin/run positive ./test/fixtures/securities/bearer_token_oas.json -b **************************
findClaims - default: Failed
  - Actual object contains a property not present in schema. Actual properties not expected: evssId, dateFiled, minEstDate, maxEstDate, waiverSubmitted, documentsNeeded, developmentLetterSent, decisionLetterSent, requestedDecision, claimType. Schema properties not found: date_filed, min_est_date, max_est_date, waiver_submitted, documents_needed, development_letter_sent, decision_letter_sent, updated_at, requested_decision, claim_type. Path: body -> data -> attributes
  - Actual value does not match schema enum. Enum values: ["\"Claim received\"","\"Initial review\"","\"Evidence gathering, review, and decision\"","\"Preparation for notification\"","\"Complete\""]. Actual value: "Complete". Path: body -> data -> attributes -> status
  - Actual object contains a property not present in schema. Actual properties not expected: evssId, dateFiled, minEstDate, maxEstDate, waiverSubmitted, documentsNeeded, developmentLetterSent, decisionLetterSent, requestedDecision, claimType. Schema properties not found: date_filed, min_est_date, max_est_date, waiver_submitted, documents_needed, development_letter_sent, decision_letter_sent, updated_at, requested_decision, claim_type. Path: body -> data -> attributes
  - Actual value does not match schema enum. Enum values: ["\"Claim received\"","\"Initial review\"","\"Evidence gathering, review, and decision\"","\"Preparation for notification\"","\"Complete\""]. Actual value: "Complete". Path: body -> data -> attributes -> status
  - Warning: This object is missing non-required parameters that were unable to be validated, including date_filed, min_est_date, max_est_date, waiver_submitted, documents_needed, development_letter_sent, decision_letter_sent, updated_at, requested_decision, claim_type. Path: body -> data -> attributes
  - Warning: This object is missing non-required parameters that were unable to be validated, including date_filed, min_est_date, max_est_date, waiver_submitted, documents_needed, development_letter_sent, decision_letter_sent, updated_at, requested_decision, claim_type. Path: body -> data -> attributes
 ›   Error: 1 operation failed
```

**Review by michaelkimball — APPROVED** [review]
lgtm

**Review by michaelkimball — APPROVED** [review]

**Pull request merged**